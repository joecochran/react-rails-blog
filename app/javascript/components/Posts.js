import React from 'react';
import axios from 'axios';
import { Link } from 'react-router-dom';
import { distanceInWordsToNow } from 'date-fns';
import { passCsrfToken } from '../util/helpers';

class Posts extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      posts: [],
    };
  }

  componentDidMount() {
    passCsrfToken(document, axios);
    this.getPosts();
    document.title = 'All Posts';
  }

  getPosts() {
    axios.get('/api/posts').then((response) => {
      this.setState({ posts: response.data.posts });
    });
  }

  deletePost(id) {
    axios.delete(`/api/posts/${id}`).then(() => {
      this.getPosts();
    });
  }

  renderAllPosts() {
    const { posts } = this.state;
    return (
      <table className="w-full text-left table-collapse table-auto">
        <thead>
          <tr>
            <th className="text-sm font-semibold text-grey-darker p-2 pl-0 bg-grey-lightest">Name</th>
            <th className="text-sm font-semibold text-grey-darker p-2 bg-grey-lightest">Category</th>
            <th className="text-sm font-semibold text-grey-darker p-2 bg-grey-lightest">Date</th>
            <th className="text-sm font-semibold text-grey-darker p-2 pr-0 bg-grey-lightest" />
          </tr>
        </thead>
        <tbody>
          {posts.map(post => (
            <tr key={post.id}>
              <td className="p-2 pl-0 border-t border-grey-light whitespace-no-wrap">
                <Link to={`/admin/posts/${post.id}`} className="text-teal-dark hover:text-teal-darker">{post.title}</Link>
              </td>
              <td className="p-2 border-t border-grey-light text-sm whitespace-no-wrap">
                {post.category.title}
              </td>
              <td className="p-2 border-t border-grey-light text-sm whitespace-no-wrap">
                <time dateTime={post.created_at}>{`${distanceInWordsToNow(post.created_at)} ago`}</time>
              </td>
              <td className="p-2 pr-0 border-t text-right border-grey-light text-xs whitespace-no-wrap">
                <button className="bg-red text-white hover:bg-red-dark font-bold rounded p-2" type="button" onClick={() => this.deletePost(post.id)}>Delete</button>
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    );
  }

  render() {
    return (
      <div className="container mx-auto">
        <h1 className="mt-6 mb-6">Posts</h1>
        {this.renderAllPosts()}
      </div>
    );
  }
}

export default Posts;
