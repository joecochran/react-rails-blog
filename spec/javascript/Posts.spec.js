import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import MockAdapter from 'axios-mock-adapter';
import Axios from 'axios';
import Posts from 'components/Posts';

configure({ adapter: new Adapter() });

// configure mock route component with call via axios
const mock = new MockAdapter(Axios);
mock.onGet('/api/posts').reply(200, {
  posts: [
    {
      id: 1,
      title: 'Foo',
      body: 'bar',
    },
  ],
});
// mock.onDelete('/api/posts/1').reply(200);

// add csrf meta tag
const meta = document.createElement('meta');
meta.setAttribute('name', 'csrf-token');
meta.content = 'fooo';
document.querySelector('head').appendChild(meta);

describe('home component', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<Posts />);
    expect(wrapper).toMatchSnapshot();
  });
});
