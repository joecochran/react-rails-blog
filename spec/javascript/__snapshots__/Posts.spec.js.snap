// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home component renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Posts />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1
          className="mt-6 mb-6"
        >
          Posts
        </h1>,
        <table
          className="w-full text-left table-collapse table-auto"
        >
          <thead>
            <tr>
              <th
                className="text-sm font-semibold text-grey-darker p-2 pl-0 bg-grey-lightest"
              >
                Name
              </th>
              <th
                className="text-sm font-semibold text-grey-darker p-2 bg-grey-lightest"
              >
                Date
              </th>
              <th
                className="text-sm font-semibold text-grey-darker p-2 pr-0 bg-grey-lightest"
              />
            </tr>
          </thead>
          <tbody />
        </table>,
      ],
      "className": "container mx-auto",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Posts",
          "className": "mt-6 mb-6",
        },
        "ref": null,
        "rendered": "Posts",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <thead>
              <tr>
                <th
                  className="text-sm font-semibold text-grey-darker p-2 pl-0 bg-grey-lightest"
                >
                  Name
                </th>
                <th
                  className="text-sm font-semibold text-grey-darker p-2 bg-grey-lightest"
                >
                  Date
                </th>
                <th
                  className="text-sm font-semibold text-grey-darker p-2 pr-0 bg-grey-lightest"
                />
              </tr>
            </thead>,
            <tbody />,
          ],
          "className": "w-full text-left table-collapse table-auto",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <tr>
                <th
                  className="text-sm font-semibold text-grey-darker p-2 pl-0 bg-grey-lightest"
                >
                  Name
                </th>
                <th
                  className="text-sm font-semibold text-grey-darker p-2 bg-grey-lightest"
                >
                  Date
                </th>
                <th
                  className="text-sm font-semibold text-grey-darker p-2 pr-0 bg-grey-lightest"
                />
              </tr>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <th
                    className="text-sm font-semibold text-grey-darker p-2 pl-0 bg-grey-lightest"
                  >
                    Name
                  </th>,
                  <th
                    className="text-sm font-semibold text-grey-darker p-2 bg-grey-lightest"
                  >
                    Date
                  </th>,
                  <th
                    className="text-sm font-semibold text-grey-darker p-2 pr-0 bg-grey-lightest"
                  />,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Name",
                    "className": "text-sm font-semibold text-grey-darker p-2 pl-0 bg-grey-lightest",
                  },
                  "ref": null,
                  "rendered": "Name",
                  "type": "th",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Date",
                    "className": "text-sm font-semibold text-grey-darker p-2 bg-grey-lightest",
                  },
                  "ref": null,
                  "rendered": "Date",
                  "type": "th",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "text-sm font-semibold text-grey-darker p-2 pr-0 bg-grey-lightest",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "th",
                },
              ],
              "type": "tr",
            },
            "type": "thead",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [],
            },
            "ref": null,
            "rendered": Array [],
            "type": "tbody",
          },
        ],
        "type": "table",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1
            className="mt-6 mb-6"
          >
            Posts
          </h1>,
          <table
            className="w-full text-left table-collapse table-auto"
          >
            <thead>
              <tr>
                <th
                  className="text-sm font-semibold text-grey-darker p-2 pl-0 bg-grey-lightest"
                >
                  Name
                </th>
                <th
                  className="text-sm font-semibold text-grey-darker p-2 bg-grey-lightest"
                >
                  Date
                </th>
                <th
                  className="text-sm font-semibold text-grey-darker p-2 pr-0 bg-grey-lightest"
                />
              </tr>
            </thead>
            <tbody />
          </table>,
        ],
        "className": "container mx-auto",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Posts",
            "className": "mt-6 mb-6",
          },
          "ref": null,
          "rendered": "Posts",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <thead>
                <tr>
                  <th
                    className="text-sm font-semibold text-grey-darker p-2 pl-0 bg-grey-lightest"
                  >
                    Name
                  </th>
                  <th
                    className="text-sm font-semibold text-grey-darker p-2 bg-grey-lightest"
                  >
                    Date
                  </th>
                  <th
                    className="text-sm font-semibold text-grey-darker p-2 pr-0 bg-grey-lightest"
                  />
                </tr>
              </thead>,
              <tbody />,
            ],
            "className": "w-full text-left table-collapse table-auto",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <tr>
                  <th
                    className="text-sm font-semibold text-grey-darker p-2 pl-0 bg-grey-lightest"
                  >
                    Name
                  </th>
                  <th
                    className="text-sm font-semibold text-grey-darker p-2 bg-grey-lightest"
                  >
                    Date
                  </th>
                  <th
                    className="text-sm font-semibold text-grey-darker p-2 pr-0 bg-grey-lightest"
                  />
                </tr>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <th
                      className="text-sm font-semibold text-grey-darker p-2 pl-0 bg-grey-lightest"
                    >
                      Name
                    </th>,
                    <th
                      className="text-sm font-semibold text-grey-darker p-2 bg-grey-lightest"
                    >
                      Date
                    </th>,
                    <th
                      className="text-sm font-semibold text-grey-darker p-2 pr-0 bg-grey-lightest"
                    />,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Name",
                      "className": "text-sm font-semibold text-grey-darker p-2 pl-0 bg-grey-lightest",
                    },
                    "ref": null,
                    "rendered": "Name",
                    "type": "th",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Date",
                      "className": "text-sm font-semibold text-grey-darker p-2 bg-grey-lightest",
                    },
                    "ref": null,
                    "rendered": "Date",
                    "type": "th",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "className": "text-sm font-semibold text-grey-darker p-2 pr-0 bg-grey-lightest",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "th",
                  },
                ],
                "type": "tr",
              },
              "type": "thead",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [],
              },
              "ref": null,
              "rendered": Array [],
              "type": "tbody",
            },
          ],
          "type": "table",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
